CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+padding or truncation semantics plus a length check beforehand: this is a cheap
+operation, unlike simulating padding semantics for example, which would have
+non-trivial cost. Secondly, due to the preponderance of truncation semantics in
+Haskell, we feel excluding it as an option is wrong. Lastly, we believe that
+outside of error checking, exact match semantics give few benefits over padding
+or truncation semantics, for performance and otherwise. This combination of
+reasoning leads us to naturally consider padding and truncation as the two to
+keep, and this guided our implementation choices.
+
+### Bit setting
+
+`builtinSetBits` in our description takes a change list argument, allowing
+changing multiple bits at once. This is an added complexity, and an argument can
+be made that something similar to the following operation would be sufficient:
+
+```haskell
+builtinSetBit :: BuiltinByteString -> BuiltinInteger -> BuiltinBool ->
+BuiltinByteString
+```
+
+Essentially, `builtinSetBit bs i v` would be equivalent to `builtinSetBits bs
+[(i, v)]` as currently defined. This was the choice made by [CIP-58][cip-58],
+with the consideration of simplicity in mind. 
+
+At the same time, due to the immutability semantics of Plutus Core, each time
+`builtinSetBit` is called, a copy of its `BuiltinByteString` argument would have
+to be made. Thus, a sequence of $k$ `builtinSetBit` calls in a fold over a
+`BuiltinByteString` of length $n$ would require $\Theta(nk)$ time and
+$\Theta(nk)$ space. Meanwhile, if we instead used `builtinSetBits`, the time
+drops to $\Theta(n + k)$ and the space to $\Theta(n)$, which is a non-trivial
+improvement. While we cannot avoid the worst-case copying behaviour of
+`builtinSetBit` (if we have a critical path of read-write dependencies of length
+$k$, for example), and 'list packing' carries some cost, we have
+[benchmarks][benchmarks-bits] that show not only that this 'packing cost' is
+essentially zero, but that for `BuiltinByteString`s of 30 bytes or fewer,
+copying completely overwhelms the work required to set the bits in the first
+place. This alone is a strong argument for having `builtinSetBits` instead;
+indeed, there is prior art for doing this [in the `vector` library][vector], for
+the exact reasons we give here.
+
+The argument could also be made whether this design should be extended to other
+primitive operations in this CIP which both take `BuiltinByteString` arguments
+and also produce `BuiltinByteString` results. We believe that this is not as
+justified as in the `builtinSetBits` case, for several reasons. Firstly, for
+`builtinLogicalComplement`, it's not even clear what benefit this would have at
+all: the only possible signature such an operation would have is
+`[BuiltinByteString] -> [BuiltinByteString]`, which in effect would be a
+specialized form of mapping. Even the _general_ form of mapping is not
+considered suitable as a primitive operation in Plutus Core! 
```

Comment:
Aside, but it's not that it's not considered suitable, but rather that we can't define it since we can't define higher-order builtin functions right now.

Reply:
Fair point, I'll amend.